tests/readerwritercircularbuffer.cpp

```diff
@@ -0,0 +1,132 @@
+{
+	BlockingReaderWriterCircularBuffer<int> q(65);
+	for (int iteration = 0; iteration != 128;
+	     ++iteration) { // check there's no problem with mismatch
+		            // between nominal and allocated capacity
+		EXPECT_EQ(q.max_capacity(), 65);
+		EXPECT_EQ(q.size_approx(), 0);
+		EXPECT_TRUE(q.try_enqueue(0));
+		EXPECT_EQ(q.max_capacity(), 65);
+		EXPECT_EQ(q.size_approx(), 1);
+		for (int i = 1; i != 65; ++i)
+			q.wait_enqueue(i);
+		EXPECT_EQ(q.size_approx(), 65);
+		EXPECT_FALSE(q.try_enqueue(65));
+
+		// Basic dequeue
+		int item;
+		EXPECT_TRUE(q.try_dequeue(item));
+		EXPECT_EQ(item, 0);
+		for (int i = 1; i != 65; ++i) {
+			q.wait_dequeue(item);
+			EXPECT_EQ(item, i);
+		}
+		EXPECT_FALSE(q.try_dequeue(item));
+		EXPECT_FALSE(q.wait_dequeue_timed(item, 1));
+		EXPECT_EQ(item, 64);
+	}
+}
+
+TEST(ReaderWriterCircularBuffer, ZeroCapacity)
+{
+	// Zero capacity
+	BlockingReaderWriterCircularBuffer<int> q(0);
+	EXPECT_EQ(q.max_capacity(), 0);
+	EXPECT_FALSE(q.try_enqueue(1));
+	EXPECT_FALSE(q.wait_enqueue_timed(1, 0));
+}
+
+void consume(BlockingReaderWriterCircularBuffer<int> *q,
+             const size_t *max_depth,
+             weak_atomic<bool> *got_mismatch)
+{
+	int item;
+	for (int i = 0; i != 1000000; ++i) {
+		EXPECT_TRUE(q->size_approx() <= *max_depth);
+		q->wait_dequeue(item);
+		if (item != i && got_mismatch->load() == false) {
+			*got_mismatch = true;
+		}
+	}
+	// printf("exiting consumer thread on item %d\n", item);
+}
+
+void produce(BlockingReaderWriterCircularBuffer<int> *q, const size_t *max_depth)
+{
+	for (int i = 0; i != 1000000; ++i) {
+		q->wait_enqueue(i);
+		EXPECT_TRUE(q->size_approx() <= *max_depth);
+	}
+	// printf("exiting producer thread\n");
+}
+
+TEST(ReaderWriterCircularBuffer, BoundedAsyncProduceAndConsume)
+{
+	// Ensure test harness checks thread-correctness on scope closure
+	testing::FLAGS_gtest_death_test_style = "threadsafe";
+
+	const size_t max_depth = 8;
+	weak_atomic<bool> got_mismatch = false;
+
+	BlockingReaderWriterCircularBuffer<int> q(max_depth);
+
+	std::thread writer(produce, &q, &max_depth);
+	std::thread reader(consume, &q, &max_depth, &got_mismatch);
+
+	writer.join();
+	reader.join();
+
+	// Make sure we've consumed all produced items and the queue is empty
+	EXPECT_EQ(q.size_approx(), 0);
+
+	// Make sure there wasn't a single out-of-sequence item consumed
+	EXPECT_FALSE(got_mismatch.load());
+
+	// Confirm both threads report as terminated
+	EXPECT_FALSE(reader.joinable());
+	EXPECT_FALSE(writer.joinable());
+
+	// When threads terminate, the OS takes time to remove them from the
+	// process's thread pool. This removal unfortunately isn't guaranteed to
+	// happen instantly (or block the main thread), so here we give the OS a
+	// bit of time to remove them. Without a bit of wait time, Google Test
+	// can sometimes catch their presence on this scope closure.
+	using namespace std::chrono_literals;
+	std::this_thread::sleep_for(100ms);
```

Comment:
I don't get it.

`.join()` is called on both threads, therefore this thread will finish only after `reader` and `writer` are done: [join()](https://en.cppreference.com/w/cpp/thread/thread/join).

There's something fishy in here or this `slepp_for` is not necessary.

Also, the comment seems misleading - there is *no* thread pool (to the best of my knowledge). `std::thread` is only a lightweight wrapper about system-specific thread implementation…

Reply:
Windows has a thread pool that doles out threads when requested; Linux has the process list. 

When I knock out the sleep, the tests pass but google adds a warning about remaining threads but mentions it can be ignored if no subsequent error occurs in the tests.  

Maybe I should have just left it, because the unit tests do pass.

It can be reproduced on its own just with two threads that simply sleep and then join. 

Maybe there are more google test calls I need similar to join for their thread checker.

Reply:
Found all the OS-specific calls where google test queries the thread pools:

https://github.com/google/googletest/blob/master/googletest/src/gtest-port.cc

Because there's no guarantee when an unlocked thread exits versus another one, then my guess is that the thread count tally (at the OS level) must also be asynchronous without strict guarantees as to when the OS actually removes the record of the thread after its join()d.

Reply:
Ok, after removing the delay, here's the warning: 

``` test
ReaderWriterCircularBuffer
[ RUN      ] ReaderWriterCircularBuffer.EnqueueDequeue
[       OK ] ReaderWriterCircularBuffer.EnqueueDequeue (12 ms)
[ RUN      ] ReaderWriterCircularBuffer.ZeroCapacity
[       OK ] ReaderWriterCircularBuffer.ZeroCapacity (0 ms)
[ RUN      ] ReaderWriterCircularBuffer.BoundedAsyncProduceAndConsume

Warning:  /build/googletest-j5yxiC/googletest-1.10.0/googletest/src/gtest-death-test.cc:1122:: Death tests use fork(), which is unsafe particularly in a threaded context. For this test, Google Test detected 2 threads. See https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#death-tests-and-threads for more explanation and suggested solutions, especially if this is the last message you see before your test times out.
[       OK ] ReaderWriterCircularBuffer.BoundedAsyncProduceAndConsume (477 ms)
[----------] 3 tests from ReaderWriterCircularBuffer (489 ms total)
```

It's likely I just misinterpreted it, because on second inspection it's not saying anything negative about our tests themselves.

The tests all pass without hanging, and it correctly detected that we had two threads.

What's your take?   Ref to CI job - https://github.com/dosbox-staging/dosbox-staging/runs/1706459350?check_suite_focus=true